File: 238. Product of Array Except Self/Solution238.py
class Solution(object):
    def productExceptSelf(self, nums):

        leftMultiplier = 1
        rightMultiplier = 1
        length = len(nums)
        ans = [0] * length
        leftArray = [0] * length
        rightArray = [0] * length
        
        for i in range(length):

            j = -i - 1
            leftArray[i] = leftMultiplier
            rightArray[j] = rightMultiplier
            leftMultiplier *= nums[i]
            rightMultiplier *= nums[j]
        
        for i in range(length):
            ans[i] = leftArray[i] * rightArray[i]
            
        return ans

c = Solution()



print(c.productExceptSelf([1,2,3,4]))
        

        

File: 238. Product of Array Except Self/Solution238.java
class Solution238 {
    public static int[] productExceptSelf(int[] nums) {

        int[] result = new int[nums.length];
        int[] leftArray = new int[nums.length];
        int[] rightArray = new int[nums.length];
        int leftSide = 1;
        int rightSide = 1;
        for (int i = 0; i < nums.length; i++) {

            int j = nums.length - 1 - i;
            leftArray[i] = leftSide;
            rightArray[j] = rightSide;
            leftSide *= nums[i];
            rightSide *= nums[j];
        }

        for (int i = 0; i < result.length; i++) {
            result[i] = leftArray[i] * rightArray[i];
        }
        return result;

    }

    public static void main(String[] args) {
        int[] nums = { 1, 2, 3, 4 };
        int[] solution = productExceptSelf(nums);
        for (int i = 0; i < solution.length; i++) {
            System.out.print(solution[i] + " ");
        }
    }
}
